[SPARK-5830][Core]Don't create unnecessary directory for local root dir

Sephiroth-Lin opened the pull request:
Now will create an unnecessary directory for local root directory, and this directory will not be deleted after application exit.
For example:
before will create tmp dir like "/tmp/spark-UUID"
now will create tmp dir like "/tmp/spark-UUID/spark-UUID"
so the dir "/tmp/spark-UUID" will not be deleted as a local root directory.

Sephiroth-Lin committed:
Don't create unnecessary directory for local root dir

Sephiroth-Lin committed:
Don't create unnecessary directory for local root dir

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
Since this appears to be exactly the same as https://issues.apache.org/jira/browse/SPARK-5801 I wonder if you can reassociate to that JIRA? and consider whether it solves the deeper nesting reported there?

srowen commented:
ok to test

SparkQA commented:
[Test build #27553 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27553/consoleFull) for   PR 4620 at commit [`26670d8`](https://github.com/apache/spark/commit/26670d83fee7c3bc0681ca775ab9f0dbc3da9d2d).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #27553 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27553/consoleFull)**     for PR 4620 at commit [`26670d8`](https://github.com/apache/spark/commit/26670d83fee7c3bc0681ca775ab9f0dbc3da9d2d)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27553/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #27560 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27560/consoleFull) for   PR 4620 at commit [`26670d8`](https://github.com/apache/spark/commit/26670d83fee7c3bc0681ca775ab9f0dbc3da9d2d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27560 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27560/consoleFull) for   PR 4620 at commit [`26670d8`](https://github.com/apache/spark/commit/26670d83fee7c3bc0681ca775ab9f0dbc3da9d2d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27560/
Test FAILed.

Sephiroth-Lin commented:
@srowen yes, this is same as SPARK-5801. In standalone, worker will create temp directories for executor, so if we create an unnecessary directory for local root directory, then when we create temp directory will create too many nested directories.

@srowen @andrewor14 from the CI report, test failed is not caused by this PR, can retest it again?

Sephiroth-Lin committed:
don't create a subdirectory for local root dir

For PR "Make sure only owner can read / write to directories created for the …" will create a subdirectory for local root dir, but this will cause nested directories, and the subdirectory for local root dir will not be deleted, which will create too many empty directories on the local root dir.

SparkQA commented:
[Test build #27604 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27604/consoleFull) for   PR 4620 at commit [`673e4e3`](https://github.com/apache/spark/commit/673e4e3c2720ef88a4316656ba9972d06d17980c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27604 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27604/consoleFull) for   PR 4620 at commit [`673e4e3`](https://github.com/apache/spark/commit/673e4e3c2720ef88a4316656ba9972d06d17980c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27604/
Test PASSed.

Sephiroth-Lin committed:
format code

SparkQA commented:
[Test build #27611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27611/consoleFull) for   PR 4620 at commit [`88e4b05`](https://github.com/apache/spark/commit/88e4b05094eb64bf4d85f54c7a5e2037bbc6f06a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27611 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27611/consoleFull) for   PR 4620 at commit [`88e4b05`](https://github.com/apache/spark/commit/88e4b05094eb64bf4d85f54c7a5e2037bbc6f06a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27611/
Test FAILed.

Sephiroth-Lin committed:
fix style

SparkQA commented:
[Test build #27623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27623/consoleFull) for   PR 4620 at commit [`c0635f1`](https://github.com/apache/spark/commit/c0635f13ce02fbf5920d9a5fb8a312fc16061b55).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27623/consoleFull) for   PR 4620 at commit [`c0635f1`](https://github.com/apache/spark/commit/c0635f13ce02fbf5920d9a5fb8a312fc16061b55).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27623/
Test PASSed.

srowen commented:
@Sephiroth-Lin yes I think this should be directed at SPARK-5801 then. SPARK-5830 is a duplicate.
CC @kayousterhout

Does this correct the many levels of extra temp dirs? it sounds like you're addressing a case where there is one extra but I probably misunderstand.

Sephiroth-Lin commented:
@srowen as in function "getOrCreateLocalRootDirs" will create a subdirectory for root local dir, then if we call "getLocalDir" will create a subdirectory for root local dir. who call getOrCreateLocalRootDirs directly. In current master branch, when we create tmp dir will call getLocaalDir first, so it will create nested directories. And in standalone mode, will create tmp dir first when lunch executor, so total it will create 4 levels directories, in other mode it will create 2 levels directories for all tmp dir.

srowen commented:
Sounds good. I'd love to get a check from @growse and/or @vanzin on this change since you can see there was a purported purpose to making the subdirectory before, which may now be moot after https://github.com/apache/spark/pull/4509

Sephiroth-Lin commented:
@srowen ok, thank you. If this subdirectory is really needed, may be we can add code to delete this subdirectory after jvm exit or sc.stop().

growse commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -686,19 +686,17 @@ private[spark] object Utils extends Logging {
+            if (rootDir.exists) {
+              Some(rootDir.getAbsolutePath)
+            } else if (rootDir.mkdirs()) {
+              chmod700(rootDir)
```
The default value of `spark.local.dir` is `/tmp`, which should exist on all linux systems at least. If I've read the patch right, that means that `getOrCreateLocalRootDirs` will simply return `/tmp`, which is usually readable by everyone. Presumably, the spark worker will then just write a load of stuff into `/tmp`. It seems that the whole purpose of the previous behaviour was to create an isolated directory inside `spark.local.dir`, because it was assumed that this value was a shared directory.

Is this not desirable behaviour, or have I missed something?

srowen commented:
Yeah, the problem was that it was creating a cascade of `.../spark-xxxx/spark-xxxx/spark-xxxx/...` directories. The desired outcome is to create exactly one `spark-xxxx` subdirectory, I think. @Sephiroth-Lin I'm not sure that's quite what this does. This would not create any subdirectories I think. Is there a way to directly and cleanly fix just the problem of extra nested dirs?

vanzin commented:
Sorry, but this patch is not correct. As @growse mentions, when `SPARK_LOCAL_DIRS` is not set, this code will try to change the permissions of `/tmp` on Unix machines. It will also use `/tmp/` as the local dir for the driver in client mode, which was the exact thing the original change was trying to avoid.

The correct fix here, if you really care about cleaning up the extra directory, is to export a different env variable from the `Worker` ([here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala#L134)) and handle that variable specially in `getOrCreateLocalRootDirs`. When that new env variable is set, the code would behave just like the `isRunningInYarnContainer()` case above the change you're making.

@srowen the current code shouldn't create a cascade of directories, but it does create a two-level-deep "spark-xxxx" hierarchy for executors in standalone mode.

vanzin commented:
Ah, wait, there's a second problem (which would result in the cascading directories, I think). `getLocalDir` should cache the local directory it returns, to avoid having to recreate it. (And should probably be made synchronized in the process.)

vanzin commented:
Hi, me again, sorry for the spam. Regarding my last comment, it's probably better if `getOrCreateLocalRootDirs()` caches its return value instead of `getLocalDir()`, since the former is called in several places, and doing that would also cover the `getLocalDir()` case.

srowen commented:
I think we should close this PR, on the grounds that the proposed change apparently goes a step too far to remove *all* nested directories. SPARK-5830 is a duplicate of SPARK-5801, which concerns creating *more than 1* nested `spark-xxx` directory. I think a PR should be reopened for SPARK-5801 that specifically addresses that issue.

Sephiroth-Lin commented:
@srowen as PR #4747 will cache the local root directories, then we can close this PR first. For PR  #4747 I think we also need to remove the local root directories after application is exited or SparkContext is stoped, or else also will create too many empty directories.

srowen commented:
Mind closing this PR?

Sephiroth-Lin commented:
@srowen ok, pls help to close this.

srowen commented:
@Sephiroth-Lin you will have to close it yourself
